core/tee/tee_sql_fs.c

```diff
@@ -0,0 +1,871 @@
+					  &ct_size, fdp->encrypted_fek);
+		if (res != TEE_SUCCESS) {
+			*errno = res;
+			goto exit;
+		}
+	}
+#else
+	copy_data(BLOCK_FILE, ct, data, BLOCK_SIZE);
+#endif
+
+	rc = sql_fs_write_rpc(fdp->fd, ct, ct_size);
+	if (rc < 0) {
+		*errno = TEE_ERROR_GENERIC;
+		goto exit;
+	}
+
+exit:
+	free(ct);
+	return rc;
+}
+
+/* Partial write (< BLOCK_SIZE) into a block: read/update/write */
+static int write_block_partial(TEE_Result *errno, struct sql_fs_fd *fdp,
+			       size_t bnum, const uint8_t *data, size_t len,
+			       size_t offset)
+{
+	size_t buf_size = BLOCK_SIZE;
+	uint8_t *buf = NULL;
+	int rc = -1;
+
+	if ((offset >= buf_size) || (offset + len > buf_size)) {
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	buf = malloc(buf_size);
+	if (!buf) {
+		*errno = TEE_ERROR_OUT_OF_MEMORY;
+		goto exit;
+	}
+
+	rc = read_block(errno, fdp, bnum, buf);
+	if (rc < 0)
+		goto exit;
+
+	memcpy(buf + offset, data, len);
+
+	rc = write_block(errno, fdp, bnum, buf);
+
+exit:
+	free(buf);
+	return rc;
+}
+
+static int sql_fs_ftruncate(TEE_Result *errno, int fd, tee_fs_off_t new_length)
+{
+	struct sql_fs_fd *fdp = handle_lookup(&fs_db, fd);
+	tee_fs_off_t old_length;
+	uint8_t *buf = NULL;
+	int rc = -1;
+
+	DMSG("(fd: %d, new_length: %" PRId64 ")...", fd, new_length);
+
+	if (!fdp) {
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit_ret;
+	}
+
+	old_length = (tee_fs_off_t)fdp->meta.length;
+
+	if (new_length == old_length) {
+		rc = 0;
+		goto exit_ret;
+	}
+
+	sql_fs_begin_transaction_rpc();
+
+	if (new_length < old_length) {
+		/* Trim unused blocks */
+		int old_last_block = block_num(old_length);
+		int last_block = block_num(new_length);
+		tee_fs_off_t off;
+
+		if (last_block < old_last_block) {
+			off = block_pos_raw(last_block);
+			rc = tee_fs_rpc_ftruncate(OPTEE_MSG_RPC_CMD_SQL_FS,
+						  fd, off);
+			if (rc < 0)
+				goto exit;
+		}
+	} else {
+		/* Extend file with zeroes */
+		tee_fs_off_t off = old_length % BLOCK_SIZE;
+		size_t bnum = block_num(old_length);
+		size_t end_bnum = block_num(new_length);
+
+		buf = calloc(1, BLOCK_SIZE);
+		if (!buf) {
+			*errno = TEE_ERROR_OUT_OF_MEMORY;
+			goto exit;
+		}
+
+		while (bnum <= end_bnum) {
+			size_t len = (size_t)BLOCK_SIZE - (size_t)off;
+
+			rc = write_block_partial(errno, fdp, bnum, buf, len,
+						 off);
+			if (rc < 0)
+				goto exit;
+			off = 0;
+			bnum++;
+		}
+	}
+
+	fdp->meta.length = new_length;
+	rc = write_meta(errno, fdp);
+
+exit:
+	sql_fs_end_transaction_rpc(rc < 0);
+	free(buf);
+exit_ret:
+	DMSG("...%d", rc);
+	return rc;
+}
+
+static tee_fs_off_t sql_fs_lseek(TEE_Result *errno, int fd,
+				 tee_fs_off_t offset, int whence)
+{
+	struct sql_fs_fd *fdp = handle_lookup(&fs_db, fd);
+	tee_fs_off_t ret = -1;
+	tee_fs_off_t raw_pos;
+	tee_fs_off_t pos;
+
+	DMSG("(fd: %d, offset: %" PRId64 ", whence: %d)...", fd, offset,
+	     whence);
+
+	if (!fdp)
+		goto exit_ret;
+
+	*errno = TEE_ERROR_GENERIC;
```

Comment:
Yes, I agree that it's always better to stick to standards and well-known ways of doing things to avoid mistakes.